---
hide:
  - navigation
---

My lab studies evolution and immunity. We're particularly interested in understanding how viruses evolve and how they interact with the human immune system. For example, how do viruses such as HIV and SARS-CoV-2 evolve to escape immunity? And how can the immune system reliably differentiate between healthy cells and ones that are cancerous or infected by intracellular pathogens? 

We develop theory and computational methods to answer these questions. The methods we develop are often inspired by statistical physics, combined with statistical learning, population genetics, and epidemiology.

<br>

![Evolution](assets/img/research/evolution.svg){ align=left }

#### Pathogen evolution

Highly mutable pathogens like HIV, influenza, and SARS-CoV-2 are a major problem for human health because they can evolve resistance to immune responses that would otherwise control infection. This is why influenza and SARS-CoV-2 vaccines need regular updates, and one of the key reasons why massive efforts have yet to produce a successful vaccine against HIV.

We aim to build quantitative models of pathogen dynamics at the level of individual hosts and populations.

**Key papers**

<small>**Parallel HIV-1 fitness landscapes shape viral dynamics in humans and macaques that develop broadly neutralizing antibodies**<br>**Shimagaki KS**, Lynch RM, **Barton JP**<br>bioRxiv [[journal link]](https://doi.org/10.1101/2024.07.12.603090) [[pdf]](assets/pdf/papers/shimagaki-parallel.pdf)  [[code]](https://github.com/bartonlab/paper-HIV-coevolution)</small>
<!-- [[si]](assets/pdf/papers/shimagaki-parallel-si.pdf) -->

<small><b>Inferring effects of mutations on SARS-CoV-2 transmission from genomic surveillance data</b><br><b>Lee B</b>, Sohail MS, <b>Finney E</b>, Ahmed SF, Quadeer AA, McKay MR<sup>c</sup>, <b>Barton JP</b><sup>c</sup><br>Nature Communications [[journal link]](https://doi.org/10.1038/s41467-024-55593-0) [[pdf]](assets/pdf/papers/lee-sc2-transmission.pdf) [[si]](assets/pdf/papers/lee-sc2-transmission-si.pdf) [[code]](https://github.com/bartonlab/paper-SARS-CoV-2-inference)</small>

<small><b>MPL resolves genetic linkage in fitness inference from complex evolutionary histories</b><br>Sohail MS<sup>=</sup>, Louie RH<sup>=</sup>, McKay MR<sup>c</sup>, <b>Barton JP</b><sup>c</sup><br>Nature Biotechnology, 2021 [[journal link]](https://doi.org/10.1038/s41587-020-0737-3) [[pdf]](assets/pdf/papers/sohail-mpl.pdf) [[si]](assets/pdf/papers/sohail-mpl-si.pdf) [[code]](https://github.com/bartonlab/paper-MPL-inference)</small>

<br>


![Immunity](assets/img/research/immunity.svg){ align=left }

#### Innate immunity

The innate immune system is our first line of defense against invading pathogens, and it also helps to fight cancer and coordinate adaptive immunity. Some innate immune cells use a diverse array of germline-encoded receptors to discriminate between healthy cells ("self") and those which are foreign or infected ("nonself"). How do innate immune cells, whose receptors are fixed and immutable, provide protection against a broad range of continually-evolving pathogens? More broadly, what principles underlie pathogen recognition in the innate immune system?

**Key papers**

<small>**An inference model gives insights into innate immune adaptation and repertoire diversity**<br>**Qin Y**, Mace EM, **Barton JP**<br>Proceedings of the National Academy of Sciences, 2023 [[journal link]](https://doi.org/10.1073/pnas.2305859120) [[pdf]](assets/pdf/papers/qin-innate.pdf) [[si]](assets/pdf/papers/qin-innate-si.pdf) [[code]](https://github.com/bartonlab/paper-innate-immune-adaptation)</small>

<br>


![Inference](assets/img/research/inference.svg){ align=left }

#### Theory and computational methods

More data than ever before is available to study evolution and immunity. Heroic sequencing efforts during the SARS-CoV-2 pandemic have made tens of millions of viral genomes available for analysis. Recently developed techniques have also made it possible to measure the functional effects of thousands of mutations in one experiment, and to quantify cellular heterogeneity in fantastic detail.

We aim to develop methods to efficiently learn from this kind of data, and to use it to predict new behavior.

**Key papers**

<small><b>Inferring effects of mutations on SARS-CoV-2 transmission from genomic surveillance data</b><br><b>Lee B</b>, Sohail MS, <b>Finney E</b>, Ahmed SF, Quadeer AA, McKay MR, <b>Barton JP</b><br>Nature Communications [[journal link]](https://doi.org/10.1038/s41467-024-55593-0) [[pdf]](assets/pdf/papers/lee-sc2-transmission.pdf) [[si]](assets/pdf/papers/lee-sc2-transmission-si.pdf) [[code]](https://github.com/bartonlab/paper-SARS-CoV-2-inference)</small>

<small>**popDMS infers mutation effects from deep mutational scanning data**<br>**Hong Z**<sup>=</sup>, **Shimagaki KS**<sup>=</sup>, **Barton JP**<br>Bioinformatics [[journal link]](https://doi.org/10.1093/bioinformatics/btae499) [[pdf]](assets/pdf/papers/hong-popdms.pdf) [[si]](assets/pdf/papers/hong-popdms-si.pdf) [[code]](https://github.com/bartonlab/popDMS)</small>

<small><b>MPL resolves genetic linkage in fitness inference from complex evolutionary histories</b><br>Sohail MS<sup>=</sup>, Louie RH<sup>=</sup>, McKay MR<sup>c</sup>, <b>Barton JP</b><sup>c</sup><br>Nature Biotechnology, 2021 [[journal link]](https://doi.org/10.1038/s41587-020-0737-3) [[pdf]](assets/pdf/papers/sohail-mpl.pdf) [[si]](assets/pdf/papers/sohail-mpl-si.pdf) [[code]](https://github.com/bartonlab/paper-MPL-inference)</small>